### ISSUE TITLE
File handle leak when starting in super dev mode

### ISSUE THREAD

**Issue by wbrophy123** [gwtproject/gwt#9433]
Starting a GWT shell in super dev mode appears to open/parse all the class files at once when loading. On larger projects, this eventually causes exceptions due to too many files being opened. 

Obviously, this could be solved by making the file handle limit much larger, but there is an underlying issue here that did not exist in the beta version of GWT 2.8.0. It appears to stem from the AnnotationConfiguration/AnnotationParser used to parse the source directories.

Command used to start GWT Shell (Obscured paths/jar list/class names):
```
/usr/java/jdk1.8.0_60/bin/java -Xmx512M -cp <JAR LIST>:/path/to/gwt/gwt-dev-linux.jar com.google.gwt.dev.DevMode -port 6983 -codeServerPort 4912 -superDevMode -bindAddress 0.0.0.0 -startupUrl <URL>.html com.company.web.gui.<namespace>.<EntryClass>
```

Here is the observed error when starting the shell:
```
2016-09-27 11:09:58.554:WARN:oejuc.AbstractLifeCycle:main: FAILED ServerConnector@6e2cd4f4{HTTP/1.1}{0.0.0.0:6983}: java.net.SocketException: Too many open files
java.net.SocketException: Too many open files
	at sun.nio.ch.Net.socket0(Native Method)
	at sun.nio.ch.Net.serverSocket(Net.java:415)
	at sun.nio.ch.ServerSocketChannelImpl.<init>(ServerSocketChannelImpl.java:88)
	at sun.nio.ch.SelectorProviderImpl.openServerSocketChannel(SelectorProviderImpl.java:56)
	at java.nio.channels.ServerSocketChannel.open(ServerSocketChannel.java:108)
	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:317)
	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:236)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.server.Server.doStart(Server.java:366)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at com.google.gwt.dev.shell.jetty.JettyLauncher.start(JettyLauncher.java:761)
	at com.google.gwt.dev.DevMode.doStartUpServer(DevMode.java:630)
	at com.google.gwt.dev.DevModeBase.startUp(DevModeBase.java:923)
	at com.google.gwt.dev.DevModeBase.run(DevModeBase.java:706)
	at com.google.gwt.dev.DevMode.main(DevMode.java:426)
2016-09-27 11:09:58.555:WARN:oejuc.AbstractLifeCycle:main: FAILED org.eclipse.jetty.server.Server@6b873080: java.net.SocketException: Too many open files
java.net.SocketException: Too many open files
	at sun.nio.ch.Net.socket0(Native Method)
	at sun.nio.ch.Net.serverSocket(Net.java:415)
	at sun.nio.ch.ServerSocketChannelImpl.<init>(ServerSocketChannelImpl.java:88)
	at sun.nio.ch.SelectorProviderImpl.openServerSocketChannel(SelectorProviderImpl.java:56)
	at java.nio.channels.ServerSocketChannel.open(ServerSocketChannel.java:108)
	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:317)
	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:236)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.server.Server.doStart(Server.java:366)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at com.google.gwt.dev.shell.jetty.JettyLauncher.start(JettyLauncher.java:761)
	at com.google.gwt.dev.DevMode.doStartUpServer(DevMode.java:630)
	at com.google.gwt.dev.DevModeBase.startUp(DevModeBase.java:923)
	at com.google.gwt.dev.DevModeBase.run(DevModeBase.java:706)
	at com.google.gwt.dev.DevMode.main(DevMode.java:426)
Unable to start embedded HTTP server
java.net.SocketException: Too many open files
	at sun.nio.ch.Net.socket0(Native Method)
	at sun.nio.ch.Net.serverSocket(Net.java:415)
	at sun.nio.ch.ServerSocketChannelImpl.<init>(ServerSocketChannelImpl.java:88)
	at sun.nio.ch.SelectorProviderImpl.openServerSocketChannel(SelectorProviderImpl.java:56)
	at java.nio.channels.ServerSocketChannel.open(ServerSocketChannel.java:108)
	at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:317)
	at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
	at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:236)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.server.Server.doStart(Server.java:366)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at com.google.gwt.dev.shell.jetty.JettyLauncher.start(JettyLauncher.java:761)
	at com.google.gwt.dev.DevMode.doStartUpServer(DevMode.java:630)
	at com.google.gwt.dev.DevModeBase.startUp(DevModeBase.java:923)
	at com.google.gwt.dev.DevModeBase.run(DevModeBase.java:706)
	at com.google.gwt.dev.DevMode.main(DevMode.java:426)

```

Using strace, I can show how these files are being opened and left open and startup (I have obscured the class names/paths). A small snippet of that output:
```
[pid   979] open("/path/to/WEB-INF/Class1$2.class", O_RDONLY) = 479
[pid   979] open("/path/to/WEB-INF/Class1$3.class", O_RDONLY) = 480
[pid   979] open("/path/to/WEB-INF/Class1$4.class", O_RDONLY) = 481
[pid   979] open("/path/to/WEB-INF/Class1$Class2.class", O_RDONLY) = 482
[pid   979] open("/path/to/WEB-INF/Class1.class", O_RDONLY) = 483
[pid   979] open("/path/to/WEB-INF/Class1$Class3$Class4.class", O_RDONLY) = 484
[pid   979] open("/path/to/WEB-INF/Class1$Class3.class", O_RDONLY) = 485
[pid   979] open("/path/to/WEB-INF/Class5$1.class", O_RDONLY) = 486
[pid   979] open("/path/to/WEB-INF/Class5$Class6.class", O_RDONLY) = 487
```
These file handles are never closed in the startup procedure.

OS in use:
````
NAME="Red Hat Enterprise Linux Server"
VERSION="7.2 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="7.2"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.2 (Maipo)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.2:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.2
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.2"
Red Hat Enterprise Linux Server release 7.2 (Maipo)
Red Hat Enterprise Linux Server release 7.2 (Maipo)
````

Hopefully that clearly explains the issue. Please let me know if there are any questions

**Comment by tbroyer** [comment]
Can you confirm that this is only about server-side classes (in `WEB-INF/`) when you run your webapp in the embedded server in DevMode; and does not happen if you run with `-noserver` or if you use the `CodeServer` entry point instead of `DevMode`?

**Comment by wbrophy123** [comment]
I cannot reproduce this error when using the -noserver option in the DevMode entry point or when starting in the CodeServer entry point. 

The classes being loaded at start up do contain client-side classes, not just server-side classes. 

Additionally, I realize now that "leak" may not be the right term to use. The files do appear to be closed, but only after the limit is reached (and then some). I then see strace go through all of those existing file handles and close all of them all at once. This only happens after it begins to fail, so it is already in a bad state at that point... It is almost as if it is closing them after some parent class begins to deallocate

**Comment by tbroyer** [comment]
If by "client-side class" you mean "in a 'client' package", but still inside `WEB-INF`, then it's a server-side class as far as the webapp is concerned. At least this hints us where to look at, but for any sufficiently big app (I.e. not a toy/weekend project) you should use a "real" servlet container, and CodeServer with `-launcherDir`, so this issue is low priority.

**Comment by jnehlmeier** [comment]
IMHO thats jetty itself that scans the class path to look for classes annotated with `@WebServlet` etc. We have also seen this in standalone Jetty in our cluster from time to time during deployment and have increased the open files limit on linux. 

I think its a general Jetty issue (which has been updated between 2.8 beta and 2.8 RC2)

Annotation scanning can be disabled using `<web-app metadata-complete="true" ...` if you do not use that feature.
